src/linear_algebra/sesquilinear_form.lean

```diff
@@ -303,4 +363,220 @@ lemma is_compl_span_singleton_orthogonal {B : V →ₗ[K] V →ₗ[K] K}
+  obtain ⟨vi, rfl⟩ := v.repr.symm.surjective m,
+  rw [basis.repr_symm_apply, finsupp.total_apply, finsupp.sum, map_sum],
+  apply finset.sum_eq_zero,
+  rintros j -,
+  rw map_smulₛₗ,
+  convert mul_zero _ using 2,
+  obtain rfl | hij := eq_or_ne i j,
+  { exact ho },
+  { exact h i j hij },
+end
+
+/-- An orthogonal basis with respect to a right-separating bilinear form has no self-orthogonal
+elements. -/
+lemma is_Ortho.not_is_ortho_basis_self_of_separating_right [nontrivial R]
+  {B : M →ₛₗ[I] M →ₛₗ[I'] R} {v : basis n R M} (h : B.is_Ortho v) (hB : B.separating_right)
+  (i : n) : ¬B.is_ortho (v i) (v i) :=
+begin
+  rw is_Ortho_flip at h,
+  rw is_ortho_flip,
+  rw separating_right_flip at hB,
+  exact h.not_is_ortho_basis_self_of_separating_left hB i,
+end
+
+/-- Given an orthogonal basis with respect to a bilinear form, the bilinear form is left-separating
+iff the basis has no elements which are self-orthogonal. -/
+lemma is_Ortho.separating_left_iff_not_is_ortho_basis_self [nontrivial R] [no_zero_divisors R]
+  {B : M →ₗ[R] M →ₗ[R] R} (v : basis n R M) (hO : B.is_Ortho v) :
+  B.separating_left ↔ ∀ i, ¬B.is_ortho (v i) (v i) :=
+begin
+  refine ⟨hO.not_is_ortho_basis_self_of_separating_left, λ ho m hB, _⟩,
+  obtain ⟨vi, rfl⟩ := v.repr.symm.surjective m,
+  rw linear_equiv.map_eq_zero_iff,
+  ext i,
+  rw [finsupp.zero_apply],
+  specialize hB (v i),
+  simp_rw [basis.repr_symm_apply, finsupp.total_apply, finsupp.sum, map_sum₂, map_smulₛₗ₂,
+    smul_eq_mul] at hB,
+  rw finset.sum_eq_single i at hB,
+  { exact eq_zero_of_ne_zero_of_mul_right_eq_zero (ho i) hB, },
+  { intros j hj hij, convert mul_zero _ using 2, exact hO j i hij, },
+  { intros hi, convert zero_mul _ using 2, exact finsupp.not_mem_support_iff.mp hi }
+end
+
+/-- Given an orthogonal basis with respect to a bilinear form, the bilinear form is right-separating
+iff the basis has no elements which are self-orthogonal. -/
+lemma is_Ortho.separating_right_iff_not_is_ortho_basis_self [nontrivial R] [no_zero_divisors R]
+  {B : M →ₗ[R] M →ₗ[R] R} (v : basis n R M) (hO : B.is_Ortho v) :
+  B.separating_right ↔ ∀ i, ¬B.is_ortho (v i) (v i) :=
+begin
+  rw is_Ortho_flip at hO,
+  rw [separating_right_flip, is_Ortho.separating_left_iff_not_is_ortho_basis_self v hO],
+  split; intros h i;
+  rw is_ortho_flip;
+  exact h i,
+end
+
+/-- Given an orthogonal basis with respect to a bilinear form, the bilinear form is nondegenerate
+iff the basis has no elements which are self-orthogonal. -/
+lemma is_Ortho.nondegenerate_iff_not_is_ortho_basis_self [nontrivial R] [no_zero_divisors R]
+  {B : M →ₗ[R] M →ₗ[R] R} (v : basis n R M) (hO : B.is_Ortho v) :
+  B.nondegenerate ↔ ∀ i, ¬B.is_ortho (v i) (v i) :=
+begin
+  refine ⟨hO.not_is_ortho_basis_self_of_separating_left ∘ and.elim_left, _⟩,
+  intros h,
+  exact ⟨(is_Ortho.separating_left_iff_not_is_ortho_basis_self v hO).mpr h,
+    (is_Ortho.separating_right_iff_not_is_ortho_basis_self v hO).mpr h⟩,
+end
```

Comment:
I am not sure whether this is good. Maybe having a lemma which proves the 'only if' for `separating_left` and then a second lemma which proves the 'iff' for `nondegenerate`?